### DISCUSSION TITLE
FC6 WriteSingleRegisterRequest Error

### DISCUSSION THREAD

**Discussion by nickblack777** [aldas/modbus-tcp-client#104] [Q&A]
Testing the WriteSingleRegisterRequest function. It gives an error when connecting:

`Packet to be sent (in hex): 64d200000006000600010037 An exception occurred
fread error during receiveFrom #0 `

**Here is my code(I have hidden the IP in setHost. He is static):**

```
require __DIR__ . '/../vendor/autoload.php';

use ModbusTcpClient\Network\BinaryStreamConnection;
use ModbusTcpClient\Packet\ModbusFunction\WriteSingleRegisterRequest;
use ModbusTcpClient\Packet\ModbusFunction\WriteSingleRegisterResponse;
use ModbusTcpClient\Packet\ResponseFactory;


$connection = BinaryStreamConnection::getBuilder()
    ->setPort(502)
    ->setHost('127.0.0.1')
    ->build();

$startAddress = 1;
$value = 55;
$packet = new WriteSingleRegisterRequest($startAddress, $value);
echo 'Packet to be sent (in hex): ' . $packet->toHex() . PHP_EOL;


try {
    $binaryData = $connection->connect()
        ->sendAndReceive($packet);
    echo 'Binary received (in hex):   ' . unpack('H*', $binaryData)[1] . PHP_EOL;

    /* @var $response WriteSingleRegisterResponse */
    $response = ResponseFactory::parseResponseOrThrow($binaryData);
    echo 'Parsed packet (in hex):     ' . $response->toHex() . PHP_EOL;
    echo 'Register value parsed from packet:' . PHP_EOL;
    print_r($response->getWord()->getInt16());

} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo '<br>';
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
} finally {
    $connection->close();
}
```

When testing via the Modbus Poll program, the value changes without errors.
https://drive.google.com/file/d/1VlH0qBCwPwIqTe2Mni437SOelCWqbX3K/view?usp=drivesdk

Please tell me what could be wrong. Thanks.

**Comment by aldas** [comment]
Are you sure that your server (where PHP is running) is allowed to connect to that modbus server and port?

**Comment by nickblack777** [comment]
Yes, I have tested receiving data via examples/index.php I can give you the ip and port of my device - please write where to send it.

**Comment by nickblack777** [comment]
Or are receiving and sending information different? I use regular virtual hosting. There were no errors while receiving the information.

**Reply by aldas** [reply]
Try writing with PHP from the same machine you are using Modbus Poll program. Your local machine and virtual hosting and modbus server can have quite different firewalls rules depending from the source/destination.

**Reply by nickblack777** [reply]
Thanks, I'll try now

**Reply by aldas** [reply]
and try to add logger to response can be seen
```php
$connection = BinaryStreamConnection::getBuilder()
    ->setPort(5020)
    ->setHost('127.0.0.1')
    ->setLogger(new EchoLogger()) // <-- add logger
    ->build();
```

**Comment by aldas** [comment]
Okey I see. From your modpoll picture I see your are using slaveID/unitID = 1 but in you code you are not setting any - thus it will be 0

try
```php
$unitID = 1;
$packet = new WriteSingleRegisterRequest($startAddress, $value, $unitID);
```

**Reply by nickblack777** [reply]
In my device, it doesn't matter what the id will be. It defaults to 0. This is not the cause of the error.

**Reply by aldas** [reply]
mmm, ok, lets go back to that picture - is your server using Modbus RTU (packets with crc) or Modbus TCP?

From that picture I see RTU packet
![image](https://user-images.githubusercontent.com/2320301/160060236-fc9a95fe-bb9e-4b32-ac10-18aa31157529.png)

if this is the case you need to convert TCP to RTU packet. See https://github.com/aldas/modbus-tcp-client/blob/master/examples/rtu.php

**Reply by nickblack777** [reply]
Modbus TCP

**Reply by nickblack777** [reply]
RTU is shown there as an example. My device is Siemens Logo. Supports Modbus TCP

**Comment by aldas** [comment]
You probably need to capture network traffic for both Modpoll and PHP packets sent/received. https://github.com/aldas/modbus-tcp-client/wiki/Debugging-guide and see what are the differences.

**Reply by nickblack777** [reply]
I think I was wrong. The Modus Poll program still uses the RTU to send changes. Although the device has TCP support. What do I need to change in order to test via RTU?

**Reply by nickblack777** [reply]
In examples found only reading of values.

**Reply by aldas** [reply]
These lines are necessary to convert TCP packet to RTU

```php
$rtuPacket = RtuConverter::toRtu($packet);
...
$response = RtuConverter::fromRtu($binaryData);
```

**Reply by nickblack777** [reply]
Same error :(

**Reply by aldas** [reply]
I think you need to capture traffic and see the differences.

The FC6 packet ifself is
https://github.com/aldas/modbus-tcp-client/blob/73eba8bdf0d4b37eda70871dc34b0c6f0659728f/src/Packet/ModbusFunction/WriteSingleRegisterRequest.php#L18-L25

In case of RTU the first 6 bytes are not included and there will be 2 extra bytes for CRC at the end.

**Reply by aldas** [reply]
You had php-fpm.log output. could you paste it again? Or increase error reporting 

etc
```php
error_repoerting(E_ALL);  
ini_set('display_errors',1);
```

**Reply by nickblack777** [reply]
These are my connection settings. All the same, the connection is TCP, not RTU. You can connect remotely from my device.

**Comment by nickblack777** [comment]
I don't know what happened, but it worked

**Comment by nickblack777** [comment]
I disabled the connection via Modbuss Poll and everything worked. I thought the devices support multiple connections at the same time. I will delete messages with my settings.

**Reply by aldas** [reply]
This is maybe configurable. I have seen WAGO PLCs that cap out at 10 tcp connections but I have seen also NMEA0183 tcp buffers/multiplexers that only allow 1 tcp connection per IP
